###**프로젝트1 load_digits**#

In [ ]:
from sklearn.datasets import load_digits, load_wine, load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, \
    confusion_matrix, recall_score, f1_score, precision_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression, SGDClassifier
from sklearn.preprocessing import StandardScaler
from matplotlib import pyplot as plt

#### Data Load & Check & Split

In [ ]:
## Data load & check

digit = load_digits()
digit.keys()
digit.target_names  # 10 class
digit.target.shape  # 1797
digit.data.shape  # (1797, 64)
digit.DESCR


".. _digits_dataset:\n\nOptical recognition of handwritten digits dataset\n--------------------------------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 5620\n    :Number of Attributes: 64\n    :Attribute Information: 8x8 image of integer pixels in the range 0..16.\n    :Missing Attribute Values: None\n    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)\n    :Date: July; 1998\n\nThis is a copy of the test set of the UCI ML hand-written digits datasets\nhttps://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits\n\nThe data set contains images of hand-written digits: 10 classes where\neach class refers to a digit.\n\nPreprocessing programs made available by NIST were used to extract\nnormalized bitmaps of handwritten digits from a preprinted form. From a\ntotal of 43 people, 30 contributed to the training set and different 13\nto the test set. 32x32 bitmaps are divided into nonoverlapping blocks of\n4x4 and the number of on pixel

In [ ]:
# Feature data & Label data Setting
data = digit.data
label = digit.target

In [ ]:
# Data Split using train_test_split
X_train, X_test, y_train, y_test = train_test_split(data, label, test_size=0.2, random_state=10)
print(X_train.shape, X_test.shape)

(1437, 64) (360, 64)


#### Model Training

In [ ]:
# Decision Tree
d_tree = DecisionTreeClassifier(random_state=10)
d_tree.fit(X_train, y_train)
tree_pred = d_tree.predict(X_test)
print(classification_report(y_test, tree_pred))

              precision    recall  f1-score   support

           0       0.94      0.92      0.93        37
           1       0.91      0.91      0.91        34
           2       0.81      0.88      0.85        34
           3       0.79      0.85      0.82        40
           4       0.73      0.79      0.76        34
           5       0.91      0.91      0.91        32
           6       0.94      0.92      0.93        37
           7       0.95      0.88      0.91        40
           8       0.74      0.70      0.72        33
           9       0.81      0.77      0.79        39

    accuracy                           0.85       360
   macro avg       0.85      0.85      0.85       360
weighted avg       0.86      0.85      0.85       360



In [ ]:
# RandomForest
d_random = RandomForestClassifier(random_state=10)
d_random.fit(X_train, y_train)
random_pred = d_random.predict(X_test)
print(classification_report(y_test, random_pred))


              precision    recall  f1-score   support

           0       1.00      0.97      0.99        37
           1       0.94      0.91      0.93        34
           2       0.97      1.00      0.99        34
           3       0.97      0.97      0.97        40
           4       0.97      0.94      0.96        34
           5       0.94      1.00      0.97        32
           6       1.00      1.00      1.00        37
           7       0.93      1.00      0.96        40
           8       0.94      0.91      0.92        33
           9       1.00      0.95      0.97        39

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360



In [ ]:
# SVM
d_svm = SVC(random_state=10)
d_svm.fit(X_train, y_train)
svm_pred = d_svm.predict(X_test)
print(classification_report(y_test, svm_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.97      1.00      0.99        34
           2       0.97      1.00      0.99        34
           3       1.00      0.97      0.99        40
           4       1.00      0.94      0.97        34
           5       0.97      1.00      0.98        32
           6       1.00      1.00      1.00        37
           7       1.00      1.00      1.00        40
           8       0.91      0.97      0.94        33
           9       1.00      0.95      0.97        39

    accuracy                           0.98       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.98      0.98      0.98       360



#### Linear Model </p>
1. error occurred. increase the number of iterations or scale the data... </p>
2. https://scikit-learn.org/stable/modules/preprocessing.html </p>
3. regarding this document, can be assumed Preprocessing data issue </p>
4. Realize that the result of Logistic Regression is binary... should do Normalize data </p>
5. choose among StandardScaler(mean 0 std 1), MinMaxScaler(min 0 max 1), RobustScaler(IQR 1) </p>


In [ ]:
# SGD Classifier
scaler = StandardScaler()
X_train_log = scaler.fit_transform(X_train)
X_test_log = scaler.fit_transform(X_test)

d_sgd = SGDClassifier(random_state=10)
d_sgd.fit(X_train_log, y_train)
sgd_pred = d_sgd.predict(X_test_log)
print(classification_report(y_test, sgd_pred))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97        37
           1       0.91      0.94      0.93        34
           2       1.00      1.00      1.00        34
           3       1.00      0.95      0.97        40
           4       0.94      0.88      0.91        34
           5       0.97      0.94      0.95        32
           6       1.00      0.97      0.99        37
           7       0.97      0.95      0.96        40
           8       0.76      0.97      0.85        33
           9       0.97      0.90      0.93        39

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



In [ ]:
# Logistic Regression
d_log = LogisticRegression(random_state=10)
d_log.fit(X_train_log, y_train)
log_pred = d_log.predict(X_test_log)
print(classification_report(y_test, log_pred))
print(confusion_matrix(y_test, svm_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        37
           1       0.92      0.97      0.94        34
           2       0.94      1.00      0.97        34
           3       1.00      0.95      0.97        40
           4       0.97      0.94      0.96        34
           5       0.91      0.97      0.94        32
           6       1.00      0.97      0.99        37
           7       1.00      1.00      1.00        40
           8       0.91      0.91      0.91        33
           9       1.00      0.95      0.97        39

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360

[[37  0  0  0  0  0  0  0  0  0]
 [ 0 34  0  0  0  0  0  0  0  0]
 [ 0  0 34  0  0  0  0  0  0  0]
 [ 0  0  1 39  0  0  0  0  0  0]
 [ 0  0  0  0 32  0  0  0  2  0]
 [ 0  0  0  0  0 32  0  0  0  0]
 [ 0  0  0  0  0  0 37  0  0  0]
 [ 0  0  0

In [ ]:
# Score Checking
print('---------Decision Tree----------')
print(d_tree.score(X_train, y_train))
print(d_tree.score(X_test, y_test))
print('---------Random Forest----------')
print(d_random.score(X_train, y_train))
print(d_random.score(X_test, y_test))
print('---------SVM----------')
print(d_svm.score(X_train, y_train))
print(d_svm.score(X_test, y_test))
print('---------SGD----------')
print(d_sgd.score(X_train_log, y_train))
print(d_sgd.score(X_test_log, y_test))
print('---------Logistic Regression----------')
print(d_log.score(X_train_log, y_train))
print(d_log.score(X_test_log, y_test))

---------Decision Tree----------
1.0
0.8527777777777777
---------Random Forest----------
1.0
0.9666666666666667
---------SVM----------
0.9944328462073765
0.9833333333333333
---------SGD----------
0.988865692414753
0.9472222222222222
---------Logistic Regression----------
0.9986082115518441
0.9666666666666667


#### Comparison of Efficiency

In [ ]:
def get_model_training(model, X_train=None, X_test=None, y_train=None, y_test=None):
    model.fit(X_train, y_train)
    predict = model.predict(X_test)
    get_eval(y_test, predict)


In [ ]:
def get_eval(y_test, predict):
    accuracy = accuracy_score(y_test, predict)
    precision = precision_score(y_test, predict, average='micro')
    recall = recall_score(y_test, predict, average='micro')
    f1 = f1_score(y_test, predict, average='micro')
    # roc_auc = roc_auc_score(y_test, predict, multi_class='ovo')
    print('Accuracy: {0:.4f}, Precision: {1:.4f}, Recall: {2:.4f}, f1: {3:.4f}'.format(accuracy, precision, recall, f1))


In [ ]:
#DT Method
model = DecisionTreeClassifier(random_state=10)
get_model_training(model, X_train, X_test, y_train, y_test)
get_eval(y_test, tree_pred)

Accuracy: 0.8528, Precision: 0.8528, Recall: 0.8528, f1: 0.8528
Accuracy: 0.8528, Precision: 0.8528, Recall: 0.8528, f1: 0.8528


#### Using for range, Get All Consequences </p>
>> Split with Linear and Non-Linear </p>

In [ ]:
model = [DecisionTreeClassifier(random_state=10), RandomForestClassifier(random_state=10), SVC(random_state=10)]
linear_model = [SGDClassifier(random_state=10), LogisticRegression(random_state=10, solver='lbfgs', max_iter=1000)]


In [ ]:
def get_model_training(model, X_train=None, X_test=None, y_train=None, y_test=None):
    for i in range(len(model)):
        model[i].fit(X_train, y_train)
        predict = model[i].predict(X_test)
        get_eval(y_test, predict)


In [ ]:
def get_eval(y_test, predict):
    accuracy = accuracy_score(y_test, predict)
    precision = precision_score(y_test, predict, average='micro')
    recall = recall_score(y_test, predict, average='micro')
    f1 = f1_score(y_test, predict, average='micro')
    # roc_auc = roc_auc_score(y_test, predict, multi_class='ovo')
    print('Accuracy: {0:.4f}, Precision: {1:.4f}, Recall: {2:.4f}, f1: {3:.4f}'.format(accuracy, precision, recall, f1))


In [ ]:
# Wanna add Naming -> 나중에 해보기
# DT, RF, SVM, SGD, LR 순서대로 값 나열
get_model_training(model, X_train, X_test, y_train, y_test)
get_model_training(linear_model, X_train_log, X_test, y_train, y_test)


Accuracy: 0.8528, Precision: 0.8528, Recall: 0.8528, f1: 0.8528
Accuracy: 0.9667, Precision: 0.9667, Recall: 0.9667, f1: 0.9667
Accuracy: 0.9833, Precision: 0.9833, Recall: 0.9833, f1: 0.9833
Accuracy: 0.8139, Precision: 0.8139, Recall: 0.8139, f1: 0.8139
Accuracy: 0.8278, Precision: 0.8278, Recall: 0.8278, f1: 0.8278


#### get_eval function Error </p>
1. ValueError: Target is multiclass but average='binary'.</p>
2. Please choose another average setting, one of [None, 'micro', 'macro', 'weighted'].</p>
3. Put 'average = ' on Precision, Recall, f1 then problem solved.</p>
4. micro, weighted are suitable, Just my thoughts.</p>
5. Most of data are not Binary i.e. Positive & Negative, so keep in mind</p>
6. https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html</p>
7. https://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html</p>
8. https://koreapy.tistory.com/506</p>
9. https://www.youtube.com/watch?v=8DbC39cvvis</p>

####  TODO </p>
1. Precision-Recall은 클래스 데이터가 불균형할 때 좋은 척도가 될 수 있음.</p>
2. Precision은 결과 관련성의 척도인반면 Recall은 얼마나 관령성 높은 결과가 반환되었는지 알 수 있는 척도</p>
3. 더 공부해봐야겠다. 위에서 두번째 website 방문.</p>
4. 임계값이란 귀무가설 하 검정통계량의 분포에서 귀무가설을 기각해야 하는 값의 집합을 정의하는 점</p>
5. ROC-AUC</p>
6. Also regarding multi classes.</p>
7. Precision & Recall Plot</p>
8. https://dojinkimm.github.io/ml/2019/10/26/handson-ch3.html (관련 설명)</p>


#### CONCLUSION

1.   DT: Accuracy: 0.8528, Precision: 0.8528, Recall: 0.8528, f1: 0.8528
2.   RF: Accuracy: 0.9667, Precision: 0.9667, Recall: 0.9667, f1: 0.9667
3.   **SVC: Accuracy: 0.9833, Precision: 0.9833, Recall: 0.9833, f1: 0.9833**
4.   SGD: Accuracy: 0.8139, Precision: 0.8139, Recall: 0.8139, f1: 0.8139
5.   LR: Accuracy: 0.8278, Precision: 0.8278, Recall: 0.8278, f1: 0.8278





현재 분류 목적은 손글씨와 digit이 맞는거니까 FP가 FN보다 중요하다 이런건 없는 것 같으니 정확도인 accuracy를 기준점으로 잡고자 함
>> SVC가 0.9833로 가장 높음

###**프로젝트2 load_wine**#

In [ ]:
wine = load_wine()
wine.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

In [ ]:
# Target Names
wine.target_names  # 3 class
wine.DESCR

'.. _wine_dataset:\n\nWine recognition dataset\n------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 178 (50 in each of three classes)\n    :Number of Attributes: 13 numeric, predictive attributes and the class\n    :Attribute Information:\n \t\t- Alcohol\n \t\t- Malic acid\n \t\t- Ash\n\t\t- Alcalinity of ash  \n \t\t- Magnesium\n\t\t- Total phenols\n \t\t- Flavanoids\n \t\t- Nonflavanoid phenols\n \t\t- Proanthocyanins\n\t\t- Color intensity\n \t\t- Hue\n \t\t- OD280/OD315 of diluted wines\n \t\t- Proline\n\n    - class:\n            - class_0\n            - class_1\n            - class_2\n\t\t\n    :Summary Statistics:\n    \n    ============================= ==== ===== ======= =====\n                                   Min   Max   Mean     SD\n    ============================= ==== ===== ======= =====\n    Alcohol:                      11.0  14.8    13.0   0.8\n    Malic Acid:                   0.74  5.80    2.34  1.12\n    Ash:                     

In [ ]:
# Data Setting
wine_data = wine.data
wine_label = wine.target

In [ ]:
# Data Split
X_train_w, X_test_w, y_train_w, y_test_w = train_test_split(wine_data, wine_label, test_size=0.2, random_state=11)


In [ ]:
# Data Normalization

scaler = StandardScaler()
X_train_wlog = scaler.fit_transform(X_train_w)
X_test_wlog = scaler.fit_transform(X_test_w)

In [ ]:
# Data Evaluation

model = [DecisionTreeClassifier(random_state=11), RandomForestClassifier(random_state=11), SVC(random_state=11)]
linear_model = [SGDClassifier(random_state=11), LogisticRegression(random_state=11, solver='lbfgs', max_iter=1000)]


In [ ]:
get_model_training(model, X_train_w, X_test_w, y_train_w, y_test_w)
get_model_training(linear_model, X_train_wlog, X_test_wlog, y_train_w, y_test_w)


Accuracy: 0.9722, Precision: 0.9722, Recall: 0.9722, f1: 0.9722
Accuracy: 0.9722, Precision: 0.9722, Recall: 0.9722, f1: 0.9722
Accuracy: 0.7778, Precision: 0.7778, Recall: 0.7778, f1: 0.7778
Accuracy: 1.0000, Precision: 1.0000, Recall: 1.0000, f1: 1.0000
Accuracy: 1.0000, Precision: 1.0000, Recall: 1.0000, f1: 1.0000


#### Conclusion
1.  wine 데이터의 분석 목적은 정확한 와인 종류 구분
2.  정확도가 중요함 
3.  linear model의 분석 accuracy가 1로 나옴 
4.  SGD Classifier & Logistic Regression 사용 
5.  이거 너무 높게 나오는데 확인해보기 

##  **프로젝트3 breast_cancer**

In [ ]:
cancer = load_breast_cancer()
cancer.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [ ]:
# Target Names
cancer.target_names  # 3 class
cancer.DESCR

'.. _breast_cancer_dataset:\n\nBreast cancer wisconsin (diagnostic) dataset\n--------------------------------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 569\n\n    :Number of Attributes: 30 numeric, predictive attributes and the class\n\n    :Attribute Information:\n        - radius (mean of distances from center to points on the perimeter)\n        - texture (standard deviation of gray-scale values)\n        - perimeter\n        - area\n        - smoothness (local variation in radius lengths)\n        - compactness (perimeter^2 / area - 1.0)\n        - concavity (severity of concave portions of the contour)\n        - concave points (number of concave portions of the contour)\n        - symmetry \n        - fractal dimension ("coastline approximation" - 1)\n\n        The mean, standard error, and "worst" or largest (mean of the three\n        largest values) of these features were computed for each image,\n        resulting in 30 features.  For instance,

In [ ]:
# Data Setting
cancer_data = cancer.data
cancer_label = cancer.target

In [ ]:
# Data Split
X_train_c, X_test_c, y_train_c, y_test_c = train_test_split(cancer_data, cancer_label, test_size=0.2, random_state=12)


In [ ]:
# Data Normalization

scaler = StandardScaler()
X_train_clog = scaler.fit_transform(X_train_c)
X_test_clog = scaler.fit_transform(X_test_c)

In [ ]:
# Data Evaluation

model = [DecisionTreeClassifier(random_state=12), RandomForestClassifier(random_state=12), SVC(random_state=12)]
linear_model = [SGDClassifier(random_state=12), LogisticRegression(random_state=12, solver='lbfgs', max_iter=1000)]


In [ ]:
get_model_training(model, X_train_c, X_test_c, y_train_c, y_test_c)
get_model_training(linear_model, X_train_clog, X_test_clog, y_train_c, y_test_c)


Accuracy: 0.9474, Precision: 0.9474, Recall: 0.9474, f1: 0.9474
Accuracy: 0.9298, Precision: 0.9298, Recall: 0.9298, f1: 0.9298
Accuracy: 0.8860, Precision: 0.8860, Recall: 0.8860, f1: 0.8860
Accuracy: 0.9561, Precision: 0.9561, Recall: 0.9561, f1: 0.9561
Accuracy: 0.9561, Precision: 0.9561, Recall: 0.9561, f1: 0.9561


#### Conclusion
1. 이 데이터는 breast cancer을 예측하는 것이기 때문에 FP보다 FN가 낮은게 중요함
2. 위양성이 나오는 위험성보다 위음성이 나오는 위험성이 더 높기 때문에
3. 즉, Recall 값을 중심적으로 보아야 한다
4. Recall값이 가장 높은 SGD Classifier & Logistic Regression 사용
5. 왜 값이 동일하지. 확인해보기